planner/core/plan_stats.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+type collectPredicateColumnsPoint struct{}
+
+func (c collectPredicateColumnsPoint) optimize(ctx context.Context, plan LogicalPlan, op *logicalOptimizeOp) (LogicalPlan, error) {
+	err := RequestLoadColumnStats(plan)
+	return plan, err
+}
+
+func (c collectPredicateColumnsPoint) name() string {
+	return "collect_predicate_columns_point"
+}
+
+type syncWaitStatsLoadPoint struct{}
+
+func (s syncWaitStatsLoadPoint) optimize(ctx context.Context, plan LogicalPlan, op *logicalOptimizeOp) (LogicalPlan, error) {
+	_, err := SyncWaitStatsLoad(plan)
+	return plan, err
+}
+
+func (s syncWaitStatsLoadPoint) name() string {
+	return "sync_wait_stats_load_point"
+}
+
+const maxDuration = 1<<63 - 1
+
+// RequestLoadColumnStats send requests to stats handle
+func RequestLoadColumnStats(plan LogicalPlan) error {
+	if plan.SCtx().GetSessionVars().InRestrictedSQL {
+		return nil
+	}
+	syncWait := plan.SCtx().GetSessionVars().StatsLoadSyncWait * time.Millisecond.Nanoseconds()
+	if syncWait <= 0 {
+		return nil
+	}
+	stmtCtx := plan.SCtx().GetSessionVars().StmtCtx
+	hintMaxExecutionTime := int64(stmtCtx.MaxExecutionTime)
+	if hintMaxExecutionTime == 0 {
+		hintMaxExecutionTime = maxDuration
+	}
+	sessMaxExecutionTime := int64(plan.SCtx().GetSessionVars().MaxExecutionTime)
+	if sessMaxExecutionTime == 0 {
+		sessMaxExecutionTime = maxDuration
+	}
+	waitTime := mathutil.MinInt64(syncWait, mathutil.MinInt64(hintMaxExecutionTime, sessMaxExecutionTime))
+	var timeout = time.Duration(waitTime)
+	_, neededColumns := CollectColumnStatsUsage(plan, false, true)
+	if config.GetGlobalConfig().Log.Level == "debug" && len(neededColumns) > 0 {
```

Comment:
`config.GetGlobalConfig().Log.Level == "debug"` is for avoiding constructing the string when the log.Level is higher?

Reply:
Yes, this block of code is only for debuging.

Reply:
Thinking about removing it, seems debug log is too much to catch information...

Reply:
removed.